activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -1012,17 +1012,18 @@ def invalidate_transaction(exception)
           return unless exception.is_a?(TransactionRollbackError)
           return unless savepoint_errors_invalidate_transactions?
 
-          current_transaction.state.invalidate! if current_transaction
+          current_transaction.state.invalidate!
+        end
+
+        def can_retry_query_on_error?(exception)
+          # We definitely can't retry if we were inside an invalidated transaction
+          return false if current_transaction.state&.invalidated?
```

Comment:
I don't particularly like the safe navigation operator here but it's needed to handle `NullTransaction` 

Perhaps design-wise it would be better for `NullTransaction` to have some kind of a `NullState` that responds to the state interface but it's a discussion for a different PR

Reply:
Didn't we have changed this to ask the state to the transaction instead of using the safe navigation in another PR? Why we are adding the same design again?

Reply:
What we added previously was `transaction#invalidate!` setter - https://github.com/rails/rails/pull/46473
We don't have state getters defined directly on the transaction object just yet. However I can add at least one for the `invalidated?` check. I decided to postpone adding getters to still have a discussion whether we want to be adding them on demand or add all of them at once or maybe still consider having a `NullTransaction` pattern